### PR TITLE
track compile success and errors

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #367 

3 new events:
```
type ProjectEvent = BaseTrack & {
    event: 'project.compiled';
    userId?: string;
    anonymousId?: string;
    properties: {
        projectType: ProjectType;
    };
};

type ProjectErrorEvent = BaseTrack & {
    event: 'project.error';
    userId?: string;
    anonymousId?: string;
    properties: {
        name: string;
        statusCode: number;
        projectType: ProjectType;
    };
};

type ApiErrorEvent = BaseTrack & {
    event: 'api.error';
    userId?: string;
    anonymousId?: string;
    properties: {
        name: string;
        statusCode: number;
        route: string;
        method: string;
    };
};
```

### CODE REVIEW

File: packages/backend/src/errors.ts [hunk 1]
```diff
@@ -171,37 +170,18 @@ export class NotFoundError extends LightdashError {
     }
 }
 
-export const errorHandler = async (error: Error, res: Response) => {
+export const errorHandler = (error: Error): LightdashError => {
```
Comment (owlas) [inline packages/backend/src/errors.ts:173]: Thanks for cleaning up!!!

**Review by TuringLovesDeathMetal — APPROVED** [review]
Looks great!

File: packages/backend/src/index.ts [hunk 2]
```diff
@@ -75,11 +76,23 @@ app.get('*', (req, res) => {
 });
 
 // errors
-app.use(
-    async (error: Error, req: Request, res: Response, next: NextFunction) => {
-        await errorHandler(error, res);
-    },
-);
+app.use((error: Error, req: Request, res: Response) => {
```
Comment (owlas) [inline packages/backend/src/index.ts:79]: Does this alter the behaviour of this middleware?

Express docs say that the ErrorHandler must be a 4 argument middleware (I guess if an error happens early then normal middleware are skipped but this should still run): https://expressjs.com/en/guide/error-handling.html (see "Writing error handlers").

Optionally: could split out the error parsing middleware (creating LightdashError) vs. the analytics middleware

**Review by owlas — COMMENTED** [review]
question about express error handlers

File: packages/backend/src/index.ts (hunk 2 above)
Comment (ZeRego) [reply packages/backend/src/index.ts:79]: nice catch :)

File: packages/backend/src/index.ts [hunk 3]
```diff
@@ -75,14 +77,32 @@ app.get('*', (req, res) => {
...
-    async (error: Error, req: Request, res: Response, next: NextFunction) => {
-        await errorHandler(error, res);
-    },
-);
+app.use((error: Error, req: Request, res: Response, next: NextFunction) => {
+    const errorResponse = errorHandler(error);
+    analytics.track({
+        event: 'api.error',
+        userId: req.user?.userUuid,
+        anonymousId: !req.user?.userUuid ? uuidv4() : undefined, // TODO: default to temporary user id once we have that
```
Comment (owlas) [inline packages/backend/src/index.ts:85]: Nice can't think of a better alternative at the moment

File: packages/backend/src/index.ts [hunk 4]
```diff
@@ -75,14 +77,32 @@ app.get('*', (req, res) => {
...
+            name: errorResponse.name,
+            message: errorResponse.message,
+            data: errorResponse.data,
+        },
+    });
+});
 
 // Update all resources on startup
-refreshAllTables().catch((e) =>
+refreshAllTables(undefined).catch((e) =>
```
Comment (owlas) [inline packages/backend/src/index.ts:105]: We'll actually drop this once we have multiple projects.

File: packages/backend/src/lightdash.ts [hunk 5]
```diff
@@ -24,32 +26,60 @@ export const getStatus = async () => {
...
-export const refreshAllTables = async () => {
+export const refreshAllTables = async (userUuid: string | undefined) => {
     tablesIsLoading = true;
     cachedTables = adapter.compileAllExplores();
     try {
         await cachedTables;
+        analytics.track({
+            event: 'project.compiled',
+            userId: userUuid,
+            anonymousId: !userUuid ? uuidv4() : undefined, // TODO: default to temporary user id once we have that
```
Comment (owlas) [inline packages/backend/src/lightdash.ts:37]: you don't need this ternary here

File: packages/backend/src/lightdash.ts [hunk 6]
```diff
@@ -24,32 +26,60 @@ export const getStatus = async () => {
...
+            properties: {
+                projectType: projectConfig.type,
+            },
+        });
+    } catch (e) {
+        const errorResponse = errorHandler(e);
+        analytics.track({
+            event: 'project.error',
+            userId: userUuid,
+            anonymousId: !userUuid ? uuidv4() : undefined, // TODO: default to temporary user id once we have that
```
Comment (owlas) [inline packages/backend/src/lightdash.ts:47]: again

**Review by owlas — COMMENTED** [review]
Just one small improvement

**Pull request merged**